### ISSUE TITLE
Advice on migrating legacy authentication to IdentityServer

### ISSUE THREAD

**Issue by RPM1984** [IdentityServer/IdentityServer3#2019]
Hi,

So i've been going through the documention/walkthroughs [here](https://identityserver.github.io/Documentation/docsv2/).

I understand the high-level concepts of OAuth/JWT's etc, now i'm trying to figure out which of the [samples](https://github.com/IdentityServer/IdentityServer3.Samples/) i should be following for my current setup, which is:

- REST API (built in Nancy). I want this to be only called by the 3 applications following. Some resources are "public" (doesn't need authentication), others are locked down to a logged in user/roles (will be handled by Claims)
- SPA (React JS)
- iOS
- Backend server proxy (MVC). Needs to call REST API (server to server).

So, from my current reading (mostly [here](https://identityserver.github.io/Documentation/docsv2/overview/simplestOAuth.html)), it appears i need to/should:

1. Create a standalone IdentityServer, hosted either standalone/IIS, or embedded in the REST API (i'd lean to former)
2. Register my API in the "scopes" of IdentityServer
3. Register my SPA application as a "client" of IdentityServer (sounds like i should be using **Implicit flow**?), with AllowedScopes = API
4. Register my iOS application as a "client" of IdentityServer (**Resource Owner** flow?), with AllowedScopes = API
5. Register my server proxy application as a "client" of IdentityServer (i guess Flow = Implicit?), with AllowedScopes = API
6. Update my API to use IdentityServer (bearer token auth, requiring API scope, check claims from token to know which resources the client can access etc)
7. Update my SPA/iOS clients to call IdentityServer to get credentials/access token and make API calls with bearer token

My questions:

1. Does the above seem reasonable? E.g what setups/flows should i have for each client?
2. How do securely get access tokens from my SPA app -> IdentityServer? If i'm passing the client credentials in order to get an access token, that means i need to store than in my SPA/JS? Not secure? (MITM etc). I'm reading that i should be using implicit flow, which doesn't support AJAX login? Bit confused about this. The other confusion i have with implicit flow, is that the user needs to supply credentials, whose claims can then be passed onto the API. But what about public/unsecure operations? I still only want my SPA/iOS app to call these, but i don't need an authenticated user.
3. How should i persist the access token in my SPA app? HTTPS-only cookie?

Basically, i'm a bit overwhelmed by all the samples/documentation and looking for more guidance, e.g "for your API, do this, for your SPA app, do this, etc".

Many thanks!

**Comment by leastprivilege** [comment]
yeah - sorry - it is a complex topic. I would recommend you start reading the OAuth2 spec and the come back with some more specific questions - also check this video:

https://vimeo.com/131635255
https://vimeo.com/113604459

**Comment by RPM1984** [comment]
@leastprivilege yeah, that's fair. Sorry.. i'll watch the videos and try my best to ask specific questions. There's just sooo many samples/options (not a bad thing!). Thanks.
